From 49bb2e8dc4015403dd84c34d73cc1cffd19f0de0 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@nierlein.de>
Date: Mon, 26 Aug 2019 15:04:17 +0200
Subject: [PATCH] fix segfault when parsing incomplete log entries (fixes #54)

parsing incomplete log entries (for example when disk runs full) may result in
segfaults because livestatus assumes the state field to be at least one byte
long.
This patch adds a check for that.

Signed-off-by: Sven Nierlein <sven@nierlein.de>
---
 src/LogEntry.cc | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/LogEntry.cc b/src/LogEntry.cc
index 9473d57..3e834c0 100644
--- a/src/LogEntry.cc
+++ b/src/LogEntry.cc
@@ -273,7 +273,12 @@ bool LogEntry::handleProgrammEntry()
 
 int LogEntry::serviceStateToInt(char *s)
 {
-    char *last = s + strlen(s) - 1;
+    char *last;
+    int len = strlen(s);
+    if(len == 0) {
+        return 3;
+    }
+    last = s + len - 1;
     if (*last == ')')
         last--;
 
@@ -291,7 +296,12 @@ int LogEntry::serviceStateToInt(char *s)
 
 int LogEntry::hostStateToInt(char *s)
 {
-    char *last = s + strlen(s) - 1;
+    char *last;
+    int len = strlen(s);
+    if(len == 0) {
+        return 3;
+    }
+    last = s + len - 1;
     if (*last == ')') // handle CUSTOM (UP) and DOWNTIMESTOPPED (DOWN)
         last--;
 
